deploy/olm-catalog/hawtio-operator/0.3.0/hawtio-operator.v0.3.0.clusterserviceversion.yaml

```diff
@@ -0,0 +1,275 @@
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  annotations:
+    alm-examples: |-
+      [{
+              "apiVersion": "hawt.io/v1alpha1",
+              "kind": "Hawtio",
+              "metadata": {
+                "name": "hawtio-online"
+              },
+              "spec": {
+                "replicas": 1,
+                "version": "1.8.0"
```

Comment:
The `version` should either be remove from the example, or valued with the digest, in offline mode.

Reply:
if replace the version with digest in UI will not show properly
![version](https://user-images.githubusercontent.com/41960152/80725583-447c0d80-8ad1-11ea-8131-d9c6841c5bb0.png)

Reply:
I already introduced, the boolean filed `UseImageDigest`- like [this](https://github.com/redhatHameed/hawtio-operator/blob/9b0b7f2da332bf703df6af31ee7fa11247b4e9bc/pkg/apis/hawtio/v1alpha1/hawtio_types.go#L32) 
so in case if it's mentioned in CR  useImageDigest: true then, it goes use the digest like [this](https://github.com/redhatHameed/hawtio-operator/blob/9b0b7f2da332bf703df6af31ee7fa11247b4e9bc/pkg/controller/hawtio/hawtio_controller.go#L903)

Reply:
If possible, I would avoid introducing a boolean in the CR for that purpose. Is there any cons detected if the version field is a digest with the `also:hash` format, aside the fact that the UI displays the digest, which is what the version is, and which would displays the same anyway with the `UseImageDigest` boolean?

Reply:
at the moment I know, the user has access to change or add incorrect value the digest value in CR, and the digest value should be the same as in CSV section related image , 
[this](https://github.com/redhatHameed/hawtio-operator/blob/9b0b7f2da332bf703df6af31ee7fa11247b4e9bc/deploy/olm-catalog/hawtio-operator/0.3.0/hawtio-operator.v0.3.0.clusterserviceversion.yaml#L266)
btw the digest retrieval also bases on tag/version of the product like [this](https://github.com/redhatHameed/hawtio-operator/blob/9b0b7f2da332bf703df6af31ee7fa11247b4e9bc/tools/constants/constants.go#L13-L15)

Reply:
I think you've lost me there :) Maybe in offline mode, the version field should not be taken into account, and the only version that can be deployed is the one specified with the digest...

Reply:
:) correct, this approached I am using for the product version for this operator, and adding an environment variable HAWTIO-IMAGE to use that image for product installation.

Reply:
I think we need to add and restrict the version of this operator like- [this](https://github.com/kiegroup/kie-cloud-operator/blob/ce1b2b36220e10ded7bdf5b9b5be5e6297a750f6/pkg/controller/kieapp/constants/constants.go#L9-L19)

Reply:
Cool, so I think we should remove the `version` field from the CSV example when it gets generation for offline mode. And warn if the user specifies it. Also, what about using a Go build variable instead of an environment variable?

Reply:
> Cool, so I think we should remove the `version` field from the CSV example when it gets generation for offline mode. And warn if the user specifies it. Also, what about using a Go build variable instead of an environment variable?

we can give both options to the user, whether to use the tag or digest, by introducing the boolean variable - to switch from offline or online.

I will look into  Go build variable how it uses and will let you know on that.

Reply:
OK, what about not introducing the boolean, and use the digest when the version is not set and the Go build variable that hold the digest has been valued. Otherwise, use `latest`?

Reply:
we can do that, but I have one concern how the user knows which product version is going to install, I think for that we need to introduce first the supported version of the product in operator [like](https://github.com/kiegroup/kie-cloud-operator/blob/ce1b2b36220e10ded7bdf5b9b5be5e6297a750f6/pkg/controller/kieapp/constants/constants.go#L9-L19). once we listed of product version, I will add these all into related image section of CSV with digest [like,](https://github.com/kiegroup/kie-cloud-operator/blob/ce1b2b36220e10ded7bdf5b9b5be5e6297a750f6/deploy/catalog_resources/redhat/7.8/businessautomation-operator.7.8.0.clusterserviceversion.yaml#L437-L512) then base version provided in CR we look into supported version digest from CSV and fetch the image.

Reply:
@astefanutti  now it's use version from CR, and get the image [like](https://github.com/redhatHameed/hawtio-operator/blob/99bcef230714d74c90ddf3361a9d9b4f9ca3fd23/pkg/controller/hawtio/hawtio_controller.go#L905) all the supported image version saved in env variable  [like](https://github.com/redhatHameed/hawtio-operator/blob/99bcef230714d74c90ddf3361a9d9b4f9ca3fd23/deploy/olm-catalog/hawtio-operator/0.3.0/hawtio-operator.v0.3.0.clusterserviceversion.yaml#L166-L172)

Reply:
@redhatHameed thanks for the update:
* Do you think we should keep the `IMAGE_REPOSITORY` environment variable?
* Is the _offline_ mode becoming the default mode for product operators, or do we need to keep support for the old _online_ mode as well?
* Do you think it'd be better to keep the versions to images mapping in a config map, or a packaged file, instead of relying on environment variables?

Reply:
> @redhatHameed thanks for the update:
> 
I also have discussed this point with my team, and here is our  the point of view
> * Do you think we should keep the `IMAGE_REPOSITORY` environment variable?

I don't think it required if we are using supported image versions in the environment variable. 

> * Is the _offline_ mode becoming the default mode for product operators, or do we need to keep support for the old _online_ mode as well?

Offline mode is not becoming the default OCP mode to our knowledge, but they need to support those users who require to install in offline mode, but using Sha's for images will work in both online AND offline mode. 

the other point is CVE addressing strategy, the common solution was to rebuild an image with the same version, That's no longer possible with the use of sha digests So if there is a major security vulnerability, there is no automated solution to address it.

in our other operator, we introduce the variable `spec.useImageTags` to switch the operator for using tags 

> * Do you think it'd be better to keep the versions to images mapping in a config map, or a packaged file, instead of relying on environment variables?

env vars is the way to go as that is the long term supported the plan for automating sha retrieval in OSBS, 
current work can be tracked here: https://projects.engineering.redhat.com/browse/CLOUDBLD-26

with that automatically do sha replacement and add a spec.relatedImages section to csv during the OSBS process.